cli/azd/pkg/infra/provisioning/bicep_provider.go

```diff
@@ -75,203 +110,350 @@ func (p *BicepProvider) Preview(ctx context.Context) *async.InteractiveTaskWithP
+	return async.RunInteractiveTaskWithProgress(
+		func(asyncContext *async.InteractiveTaskContextWithProgress[*DeployResult, *DeployProgress]) {
+			done := make(chan bool)
+			var operations []azcli.AzCliResourceOperation
 
-	parameters := make(map[string]BicepInputParameter)
+			deploymentSlug := azure.SubscriptionDeploymentRID(p.env.GetSubscriptionId(), p.env.GetEnvName())
+			deploymentUrl := fmt.Sprintf(output.WithLinkFormat("https://portal.azure.com/#blade/HubsExtension/DeploymentDetailsBlade/overview/id/%s\n"), url.PathEscape(deploymentSlug))
+			p.console.Message(ctx, fmt.Sprintf("Provisioning Azure resources can take some time.\n\nYou can view detailed progress in the Azure Portal:\n%s", deploymentUrl))
 
-	for key, param := range preview.Parameters {
-		parameters[key] = BicepInputParameter(param)
-	}
+			// Ensure the done marker channel is sent in all conditions
+			defer func() {
+				done <- true
+			}()
 
-	bicepFile.Parameters = parameters
+			// Report incremental progress
+			go func() {
+				resourceManager := infra.NewAzureResourceManager(ctx)
+				progressDisplay := NewProvisioningProgressDisplay(resourceManager, p.console, p.env.GetSubscriptionId(), p.env.GetEnvName())
+
+				for {
+					select {
+					case <-done:
+						return
+					case <-time.After(10 * time.Second):
+						progressReport, err := progressDisplay.ReportProgress(ctx)
+						if err != nil {
+							// We don't want to fail the whole deployment if a progress reporting error occurs
+							log.Printf("error while reporting progress: %s", err.Error())
+							continue
+						}
+
+						operations = progressReport.Operations
+						asyncContext.SetProgress(progressReport)
+					}
+				}
+			}()
 
-	bytes, err := json.MarshalIndent(bicepFile, "", "  ")
-	if err != nil {
-		return fmt.Errorf("marshaling parameters: %w", err)
-	}
+			// Start the deployment
+			modulePath := p.modulePath()
+			parametersFilePath := p.parametersFilePath()
+			deployResult, err := p.deployModule(ctx, scope, modulePath, parametersFilePath)
 
-	parametersFilePath := p.parametersFilePath()
-	err = os.WriteFile(parametersFilePath, bytes, 0644)
-	if err != nil {
-		return fmt.Errorf("writing parameters file: %w", err)
-	}
+			if err != nil {
+				asyncContext.SetError(err)
+				return
+			}
 
-	return nil
-}
+			if deployResult != nil {
+				deployment.Outputs = p.createOutputParameters(deployment, deployResult.Properties.Outputs)
+			}
 
-// Provisioning the infrastructure within the specified template
-func (p *BicepProvider) Deploy(ctx context.Context, preview *Preview, scope Scope) *async.InteractiveTaskWithProgress[*DeployResult, *DeployProgress] {
-	return async.RunInteractiveTaskWithProgress(
-		func(asyncContext *async.InteractiveTaskContextWithProgress[*DeployResult, *DeployProgress]) {
-			isDeploymentComplete := false
+			result := &DeployResult{
+				Operations: operations,
```

Comment:
There's a possible race condition here. `operations` is a shared resource. The display goroutine can still run since we haven't send the `done` message yet. So the simplest fix is to move `done` prior to using `operations`.

That still leaves possibilities for "future us" to use `operations` prior to sending `done`, which I think the cleanest way is to use channels to communicate the result.

Reply:
This is a tricky one as well because there is still a large window in which done can happen during the 10 second timeout.  I haven't see any instances in my testing where this was the case.  Worst case scenario is you have a slightly outdated version of the operations return as part of the result.  

Lets address with something more robust in the future if we see issues.

Reply:
It's possible to get a segfault if we happen to increase the array size and had to reallocate memory for the array. It's very unlikely to happen, but based on the code, there's a possibility. Can we move `done` to right after `deployModule`? I think that prevents it for now, since we guarantee no writers after waiting for the goroutine to tear down.

Note: with or without this change, we have to wait for the goroutine to tear down. This means waiting for a UI refresh operation to be done if the deploy finished while the UI was refreshing.